### DISCUSSION TITLE
How to make colspec allow variables?

### DISCUSSION THREAD

**Discussion by note286** [lvjr/tabularray#263] [Q&A]
My requirement is how to get colspec to receive variable values. Here is my exploration process.

```latex
\documentclass{article}
\usepackage{tabularray}
\begin{document}
\begin{tblr}{colspec={r|X}}
AAAAAA & Lorem\ ipsum\ dolor\ sit\ amet,\ consectetur\ adipisicing\ elit,\ sed\ do\ eiusmod\ tempor \\
AAA & BBB \\
\end{tblr}
\end{document}
```
![image](https://user-images.githubusercontent.com/16207292/172384657-52ca7967-7329-42d7-948a-120a65ee5270.png)

As shown above, everything is normal.

I found the function `\__tblr_parse_colrow_spec:nn { column } { }` by reading the source code, and I tried to configure the column options by calling the function `\__tblr_parse_colrow_spec:nn { column } { }` directly.

```latex
\documentclass{article}
\usepackage{tabularray}
\begin{document}
\ExplSyntaxOn
\begin{tblr}{}
\__tblr_parse_colrow_spec:nn { column } { r|X }
AAAAAA & Lorem\ ipsum\ dolor\ sit\ amet,\ consectetur\ adipisicing\ elit,\ sed\ do\ eiusmod\ tempor \\
AAA & BBB \\
\end{tblr}
\ExplSyntaxOff
\end{document}
```

But I got the following results. The vertical bar is not displayed correctly. What causes this?

![image](https://user-images.githubusercontent.com/16207292/172385740-647e4594-839e-4e3a-89fc-62016991364c.png)

Then, I directly modify the `tabularray.sty` file to generate variant function to the function `\__tblr_parse_colrow_spec:nn` so as to receive variables.

```latex
\cs_generate_variant:Nn \__tblr_parse_colrow_spec:nn { nV }
```

And modified the function corresponding to colspec parameter.

```latex
% colspec .code:n = \__tblr_parse_colrow_spec:nn { column } {#1},
colspec .code:n = \__tblr_parse_colrow_spec:nV { column } {#1},
```

Then we run the following example:

```latex
\documentclass{article}
\usepackage{tabularray}
\begin{document}
\ExplSyntaxOn
\tl_new:N \l_tmp_colspec_tl
\tl_set:Nn \l_tmp_colspec_tl { r|X }
\begin{tblr}{colspec=\l_tmp_colspec_tl}
AAAAAA & Lorem\ ipsum\ dolor\ sit\ amet,\ consectetur\ adipisicing\ elit,\ sed\ do\ eiusmod\ tempor \\
AAA & BBB \\
\end{tblr}
\ExplSyntaxOff
\end{document}
```

We got the expected result.

![image](https://user-images.githubusercontent.com/16207292/172387892-b85f9494-9662-42db-be20-ce36b1b5ff85.png)

In conclusion, my purpose is to use variables to configure column options. If the first example works correctly, then I can pass the variable to the variant function without modifying the `.sty` file. At present, I can't find a suitable way to meet my needs.

**Comment by lvjr** [comment]
It is better to define a new environment, because you can do anything you want without modifying private functions in the package:

```latex
\documentclass{article}
\usepackage{tabularray}
\begin{document}
\ExplSyntaxOn

\tl_new:N \l__mytest_keyval_tl

\keys_define:nn {mytest}
  {
    colspec .code:n = \__mytest_store_keyval:no { colspec } {#1},
    unknown .code:n = \__mytest_store_keyval:Vn \l_keys_key_str {#1},
  }
\cs_new_protected:Npn \__mytest_store_keyval:nn #1 #2
  {
    \tl_if_empty:nTF {#2}
      { \tl_put_right:Nn \l__mytest_keyval_tl {#1,} }
      { \tl_put_right:Nn \l__mytest_keyval_tl {#1 = {#2},} }      
  }
\cs_generate_variant:Nn \__mytest_store_keyval:nn {no,Vn}

\NewDocumentEnvironment {mytblr} {O{} m +b}
  {
    \tl_clear:N \l__mytest_keyval_tl
    \keys_set:nn {mytest} {#2}
    \tl_set:Nx \l_tmpa_tl
      {
        \exp_not:n {\begin{tblr}[#1]} {\exp_not:V \l__mytest_keyval_tl}
          \exp_not:n {#3}
        \exp_not:n {\end{tblr}}
      }
    \tl_use:N \l_tmpa_tl
  } {}

\tl_new:N \l_tmp_colspec_tl
\tl_set:Nn \l_tmp_colspec_tl { r|X }

\begin{mytblr}{colspec=\l_tmp_colspec_tl,hlines,rowsep=1pt}
AAAAAA & Lorem\ ipsum\ dolor\ sit\ amet,\ consectetur\ adipisicing\ elit,\ sed\ do\ eiusmod\ tempor \\
AAA & BBB \\
\end{mytblr}

\ExplSyntaxOff
\end{document}
```

![image](https://user-images.githubusercontent.com/4239949/172526140-16c38ee1-0970-4cd9-933f-ad05e60df094.png)

**Reply by note286** [reply]
Thank you very much for your answer, which completely solved my problem.
In addition, I would like to ask why direct use of `\__tblr_parse_colrow_spec:nn { column } { r|X }` can't get the expected effect? As shown in the second example in the question.

**Reply by lvjr** [reply]
From the manual: "All commands which change the specifications of tables **must** be defined with `\NewTableCommand`.
```latex
\documentclass{article}
\usepackage{tabularray}
\begin{document}
\ExplSyntaxOn
\NewTableCommand \mycolspec [1] {\__tblr_parse_colrow_spec:nn { column } { #1 } }
\begin{tblr}{}
\mycolspec{ r|X }
AAAAAA & Lorem\ ipsum\ dolor\ sit\ amet,\ consectetur\ adipisicing\ elit,\ sed\ do\ eiusmod\ tempor \\
AAA & BBB \\
\end{tblr}
\ExplSyntaxOff
\end{document}
```
![image](https://user-images.githubusercontent.com/4239949/172801289-289fbaad-909e-4911-add3-1c362412a2b7.png)

But it is always dangerous to use private functions such as `\__tblr_parse_colrow_spec:nn` in your package. If you use them the documents of your users may break down in the future. Therefore it is better to keep your package away from them.

**Reply by note286** [reply]
Thank you very much for your answer.
